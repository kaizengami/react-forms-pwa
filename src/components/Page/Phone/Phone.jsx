import { h, render, Component } from "preact";
import { route } from "preact-router";

import Error from "../../Error/Error";

class Phone extends Component {
  constructor(props) {
    super(props);

    this.state = {
      active: false,
      error: false
    };
    this.submit = this.submit.bind(this);
    this.onTyping = this.onTyping.bind(this);
    this.outOfFocus = this.outOfFocus.bind(this);
  }

  submit(e) {
    e.preventDefault();
    let value = this.input.value;
    if (value.length < 6) {
      //this.error.innerHTML = "not valid number";
      this.setState({ error: true });
    } else {
      this.setState({ error: false });
      route("/name");
    }
    this.setState({ error: value.length < 6 });
  }

  isKeyNumber(char) {
    if (/^[0-9]*$/.test(char)) {
      return true;
    }
    return false;
  }

  onTyping() {
    const inputValue = this.input.value;
    const mask = this.input.getAttribute("mask");
    const lastChar = inputValue.slice(-1);

    if (this.isKeyNumber(lastChar)) {
      this.applyMask(inputValue, mask);
    } else {
      this.input.value = inputValue.slice(0, -1);
    }
  }

  applyMask(inputValue, mask) {
    let coutryCode = mask.match(/{([^}]+)}/)[1];
    let hyphenIndexes = this.getHyphenPosition(mask);

    if (inputValue.includes(coutryCode)) {
      inputValue = inputValue.replace(/-/g, "");
      let stringArray = inputValue.split("");
      hyphenIndexes.forEach((index, i) => {
        if (inputValue.length < index) return;
        stringArray.splice(index, 0, "-");
      });
      this.input.value = stringArray.join("");
    } else this.input.value = coutryCode;
  }

  getHyphenPosition(mask) {
    let indexes = [];
    let string = mask.replace(/{/, "").replace(/}/, "");
    for (let i = 0; i < string.length; i++) {
      if (string.charAt(i) === "-") {
        indexes.push(i);
        // const test = [...indexes, i]
      }
    }
    return indexes;
  }

  outOfFocus() {
    if (this.input.value.length === 4) this.input.value = "";
  }

  componentDidMount() {
    this.props.setStep(1);
  }

  render() {
    return (
      <div className="phone-wrapper">
        <form className="phone-form" onSubmit={this.submit}>
          <div className="phone-title">Phone number:</div>
          <input
            className="phone-input"
            type="tel"
            name="number"
            autoComplete="off"
            ref={input => (this.input = input)}
            data-testid="input"
            onKeyDown={this.isKeyNumber}
            onKeyUp={this.onTyping}
            onFocus={this.onTyping}
            onBlur={this.outOfFocus}
            mask="{+234}-000-000-000"
          />
          <input
            className="phone-input-hidden"
            type="submit"
            id="submit-number"
          />
        </form>
        <div
          className={
            this.state.error ? "phone-error-box" : "phone-error-box-hide"
          }
        >
          <Error error="phone number" />
        </div>
        <label
          className="phone-submit-label"
          htmlFor="submit-number"
          tabIndex={0}
        >
          Submit
        </label>
      </div>
    );
  }
}

export default Phone;
