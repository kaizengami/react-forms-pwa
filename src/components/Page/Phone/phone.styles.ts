import styled from "styled-components";

export const PhoneWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
  align-items: center;
`;

export const Title = styled.div`
  font-size: 30px;
`;

export const Form = styled.form`
  box-shadow: 0 2px 4px 0 rgba(0, 0, 0, 0.2), 0 3px 10px 0 rgba(0, 0, 0, 0.19);
  width: 300px;
  padding: 20px 30px;
`;

export const Input = styled.input`
  font-size: 24px;
  height: 40px;
  border: 0;
  outline: 0;
  background: transparent;
  border-bottom: 1px solid #00000040;
  ::-webkit-inner-spin-button,
  ::-webkit-outer-spin-button {
    -webkit-appearance: none;
    margin: 0;
  }
`;

export const SubmitHidden = styled.input`
  display: none;
`;

export const Submit = styled.label`
  font-size: 18px;
  width: 120px;
  height: 40px;
  color: #ffffff;
  text-align: center;
  line-height: 40px;
  border: none;
  margin-top: 50px;
  border-radius: 3px;
  margin-left: 240px;
  background-image: -webkit-linear-gradient(
    0deg,
    rgb(242, 113, 33) 0%,
    rgb(233, 64, 87) 50%,
    rgb(138, 35, 135) 100%
  );
  box-shadow: 0px 3px 10px 0px rgba(0, 0, 0, 0.15);
  cursor: pointer;
`;

export const Error = styled<any>("div")`
  position: absolute;
  margin-top: 130px;
  display: ${props => (props.error ? "block" : "none")};
  color: red;
`;
