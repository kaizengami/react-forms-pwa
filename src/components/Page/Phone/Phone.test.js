import React from "react";
import { create } from "react-test-renderer";

import Phone from "./Phone";

import { render, fireEvent } from "react-testing-library";
import "react-testing-library/cleanup-after-each";
import "jest-dom/extend-expect";

it("renders the component Phone", () => {
  const container = render(<Phone />);
  expect(container.firstChild).toMatchSnapshot();
});

describe("Submit", () => {
  it("save text", async () => {
    const { getByText, getByTestId } = render(<Phone />);
    const input = getByTestId("input");
    input.value = "123";
    fireEvent.change(input);
    fireEvent.click(getByText("Submit"));
    console.log("Submit", getByTestId("error").innerHTML);
    expect(getByTestId("error")).toHaveTextContent("not valid number");
  });
});
