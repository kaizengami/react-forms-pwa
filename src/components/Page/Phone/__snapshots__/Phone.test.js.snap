// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the component Phone 1`] = `undefined`;
